### ISSUE TITLE
`tree_map` all operations in a single optical plane

### ISSUE THREAD

**Issue by LouisDesdoigts** [LouisDesdoigts/dLux#155]
So this might be a bit tedious to do, but it could be possible to vectorise multiple operations within a single (physical) optical plane using `jax.tree_map`. At init time in the `Optics` class, we could run through the layers and chunk operations together between specified layers that can be performed simultaneously. As an example, say we have the following set up:

```python
layers = [
    CreateWavefront(),    # Block 1
    ApplyBasisOPD(),      # Block 1
    TransmissiveOptic(),  # Block 1
    NormaliseWavefront(),  # Block 2
    TiltWavefront(),       # Block 2
    AddOPD(),              # Block 2
    AngularMFT(),           # Block 3
    TranmissiveOptic(),      # Block 4
    ApplyOPD(),              # Block 4
    AngularMFT(),             # Block 5
    TransmissiveOptic(),       # Block 6
    AngularMFT()                # Block 7
]
```

Each operation within these blocks can be performed simultaneously, outputting an amplitude and phase that can then be applied to the wavefront. There are two speed improvements we get here:
 1. Simultaneous calculations means we can get better utilisation of hardware at large scales, and faster low-chromaticity calculation that aren't able to fully utilise the hardware available currently.
 2. Reduces the number of operations on the wavefront specifically, lowering the amount of re-writing of the objects (since it is immutable).

This Is unlikely to provide large speed increases in most cases, but could be an quite effective in some scenarios.

There are some thorny questions however, for example the wavefront normalisation depends on the order in which the amplitude modifying layers are applied. Issue #154 discusses a possible improvement that might make this simpler. For the most part however these blocks can be simply generated by using the propagator layers to splitting the blocks

**Comment by LouisDesdoigts** [comment]
This is essentially just a trade off of vectorisation vs memory usage. Ultimately we can always wait longer, but we cant necessarily add more ram, so this will not be implemented.
